examples/payment_channel/lib.rs

```diff
@@ -0,0 +1,475 @@
+            // Emit event
+            self.env().emit_event(SenderCloseStarted {
+                expiration: now + self.close_duration,
+                close_duration: self.close_duration,
+            });
+
+            self.expiration = expiration;
+        }
+
+        // If the timeout is reached without the recipient closing the channel, then
+        // the token is released back to the sender.
+        #[ink(message)]
+        pub fn claim_timeout(&mut self) {
+            self.ensure_timeout();
+            self.env().terminate_contract(self.sender);
+        }
+
+        // Recipient can withdraw the funds from the channel at any time.
+        #[ink(message)]
+        pub fn withdraw(&mut self, amount: Balance, signature: [u8; 65]) {
+            self.ensure_recipient();
+
+            // Signature validation
+            self.ensure_valid_signature(amount, signature);
+
+            // Make sure there's something to withdraw (guards against underflow)
+            assert!(
+                amount > self.withdrawn,
+                "amount must be greater than withdrawn"
+            );
+            let amount_to_withdraw = amount - self.withdrawn;
+            self.withdrawn += amount_to_withdraw;
+
+            if self
+                .env()
+                .transfer(self.recipient, amount_to_withdraw)
+                .is_err()
+            {
+                panic!(
+					"requested transfer failed. this can be the case if the contract does not\
+                     have sufficient free funds or if the transfer would have brought the\
+                     contract's balance below minimum balance."
+				)
+            }
+        }
+
+        #[ink(message)]
+        pub fn get_sender(&self) -> AccountId {
+            self.sender
+        }
+
+        #[ink(message)]
+        pub fn get_recipient(&self) -> AccountId {
+            self.recipient
+        }
+
+        #[ink(message)]
+        pub fn get_expiration(&self) -> Timestamp {
+            self.expiration
+        }
+
+        #[ink(message)]
+        pub fn get_withdrawn(&self) -> Balance {
+            self.withdrawn
+        }
+
+        #[ink(message)]
+        pub fn get_close_duration(&self) -> Timestamp {
+            self.close_duration
+        }
+
+        #[ink(message)]
+        pub fn get_account_id(&self) -> AccountId {
+            self.env().account_id()
+        }
+
+        #[ink(message)]
+        pub fn get_balance(&self) -> Balance {
+            self.env().balance()
+        }
+    }
+
+    #[ink(impl)]
+    impl PaymentChannel {
+        fn ensure_sender(&self) {
+            assert!(self.env().caller() == self.sender, "not the sender");
+        }
+
+        fn ensure_recipient(&self) {
+            assert!(self.env().caller() == self.recipient, "not the recipient");
+        }
+
+        fn ensure_timeout(&self) {
+            let now = self.env().block_timestamp();
+            assert!(now >= self.expiration, "not timed out yet");
+        }
```

Comment:
These are only used once or twice, I'd just inline them

Reply:
It's done!